pkg/workerpool/pool_test.go

```diff
@@ -368,3 +368,52 @@ func (s *workerPoolSuite) TestBasics(c *check.C) {
 	err := errg.Wait()
 	c.Assert(err, check.ErrorMatches, "context canceled")
 }
+
+// TestCancelByAddEventContext makes sure that the event handle can be cancelled by the context used
+// to call `AddEvent`.
+func (s *workerPoolSuite) TestCancelByAddEventContext(c *check.C) {
+	defer testleak.AfterTest(c)()
+
+	poolCtx, poolCancel := context.WithCancel(context.Background())
+	defer poolCancel()
+	pool := newDefaultPoolImpl(&defaultHasher{}, 4)
+	go func() {
+		err := pool.Run(poolCtx)
+		c.Assert(err, check.ErrorMatches, ".*context canceled.*")
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*20)
+	defer cancel()
+	errg, ctx := errgroup.WithContext(ctx)
+
+	for i := 0; i < 8; i++ {
+		handler := pool.RegisterEvent(func(ctx context.Context, event interface{}) error {
+			<-ctx.Done()
+			return ctx.Err()
+		})
+
+		errg.Go(func() error {
+			for j := 0; j < 64; j++ {
+				err := handler.AddEvent(ctx, j)
+				if err != nil {
+					return nil
+				}
```

Comment:
if return `nil` here, should add an assertion to the error?

Reply:
No. `AddEvent` is allowed to return an error if `ctx` has already been cancelled, which will happen here.

Reply:
Yes, `AddEvent` could return `context.Canceled` or `ErrWorkerPoolHandleCancelled`, both these two errors are expected?

Reply:
We are not test cancelling the handles, so `ErrWorkerPoolHandleCancelled` is not expected in this test case.